rust/lance-encoding/src/encodings/physical/binary.rs

```diff
@@ -457,6 +462,220 @@ fn get_indices_from_string_arrays(
     (indices, null_adjustment)
 }
 
+#[derive(Debug, Default)]
+pub struct BinaryMiniBlockEncoder {}
+
+const AIM_MINICHUNK_SIZE: u64 = 4 * 1024;
+
+// search for the next offset index to cut the values into a chunk.
+// this function incrementally peek the number of values in a chunk,
+// each time multiplies the number of values by 2.
+// It returns the offset_idx in `offsets` that belongs to this chunk.
+fn search_next_offset_idx(offsets: &[u32], last_offset_idx: usize) -> usize {
+    let mut num_values = 1;
+    loop {
+        if last_offset_idx + num_values >= offsets.len() {
+            return offsets.len() - 1;
+        }
+        if ((offsets[last_offset_idx + num_values] - offsets[last_offset_idx])
+            + ((num_values + 1) * 4) as u32)
+            <= AIM_MINICHUNK_SIZE as u32
+        {
+            num_values *= 2;
+        } else {
+            break;
+        }
+    }
+    last_offset_idx + (num_values >> 1)
+}
+
+impl BinaryMiniBlockEncoder {
+    // put binary data into chunks, every chunk is less than or equal to `AIM_MINICHUNK_SIZE`.
+    // In each chunk, offsets are put first then followed by binary bytes data, each chunk is padded to 8 bytes. 
+    // the offsets in the chunk points to the bytes offset in this chunk.
+    fn chunk_data(
+        &self,
+        mut data: VariableWidthBlock,
+    ) -> (MiniBlockCompressed, crate::format::pb::ArrayEncoding) {
+        let max_len = data
+            .get_stat(Stat::MaxLength)
+            .expect("VariableWidthDataBlock should have valid max length statistics");
+        let max_len = max_len
+            .as_any()
+            .downcast_ref::<PrimitiveArray<UInt64Type>>()
+            .unwrap();
+
+        assert!(max_len.value(0) < 128);
+        assert!(data.bits_per_offset == 32);
+
+        let offsets = data.offsets.borrow_to_typed_slice::<u32>();
+        let offsets = offsets.as_ref();
+
+        assert!(offsets.len() > 1);
+
+        #[derive(Debug)]
+        struct ChunkInfo {
+            chunk_start_offset_in_orig_idx: usize,
+            chunk_last_offset_in_orig_idx: usize,
+            // the bytes in every chunk starts at `chunk.bytes_start_offset`
+            bytes_start_offset: usize,
+            // every chunk is padded to 8 bytes.
+            // we need to interpret every chunk as &[u32] so we need it to padded at least to 4 bytes,
+            // 8 bytes are a more conservative behavior.
```

Comment:
I think there is some confusing between padding and alignment here.  You want to be able to take the offsets of a chunk and interpret them as `&[u32]`.  In order for this to work the slice needs to be _aligned to_ (not padded) 4 bytes (and it is quite concretely 4, let's not make it 8).  The means the pointer to the start of the slice needs to be divisible by 4.

Padding the buffer to make the length divisible by 4 does not guarantee this, except possibly accidentally.

In fact, getting this alignment correct is kind of tricky, and is going to take some tweaking of the miniblock encoder itself :frowning:.

Let me throw together a PR.  We can probably change our "pad to 6" logic to just be "pad to 8" and then pad the start of the data buffer appropriately so that the start of the data buffer is on an 8-byte aligned pointer.  This will ensure that each chunk starts with an 8-byte aligned pointer.

Reply:
I did a copy of the entire chunk in `decompress`(hope we can eliminate it after your PR). 

because I did a copy, I think I can safely reinterpret the chunk as `&[u32]` as long as it's length in `number of bytes` is some multiple of 4.